// html2canvas-pro@2.2.0/dist/lib/render/effects.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isFilterEffect = exports.isBlendEffect = exports.isClipPathEffect = exports.isOpacityEffect = exports.isClipEffect = exports.isTransformEffect = exports.FilterEffect = exports.BlendEffect = exports.ClipPathEffect = exports.OpacityEffect = exports.ClipEffect = exports.TransformEffect = void 0;
const mix_blend_mode_1 = require("../css/property-descriptors/mix-blend-mode");
class TransformEffect {
    constructor(offsetX, offsetY, matrix) {
        this.offsetX = offsetX;
        this.offsetY = offsetY;
        this.matrix = matrix;
        this.type = 0 /* EffectType.TRANSFORM */;
        this.target = 2 /* EffectTarget.BACKGROUND_BORDERS */ | 4 /* EffectTarget.CONTENT */;
    }
}
exports.TransformEffect = TransformEffect;
class ClipEffect {
    constructor(path, target) {
        this.path = path;
        this.target = target;
        this.type = 1 /* EffectType.CLIP */;
    }
}
exports.ClipEffect = ClipEffect;
class OpacityEffect {
    constructor(opacity) {
        this.opacity = opacity;
        this.type = 2 /* EffectType.OPACITY */;
        this.target = 2 /* EffectTarget.BACKGROUND_BORDERS */ | 4 /* EffectTarget.CONTENT */;
    }
}
exports.OpacityEffect = OpacityEffect;
/**
 * Clips the element and all its descendants to an arbitrary canvas-drawn shape.
 * The `applyClip` callback is responsible for calling beginPath, the shape
 * operations, and ctx.clip() — giving each shape type full control over how
 * the path is constructed (arc, ellipse, lineTo, Path2D, etc.).
 */
class ClipPathEffect {
    constructor(applyClip) {
        this.applyClip = applyClip;
        this.type = 3 /* EffectType.CLIP_PATH */;
        this.target = 2 /* EffectTarget.BACKGROUND_BORDERS */ | 4 /* EffectTarget.CONTENT */;
    }
}
exports.ClipPathEffect = ClipPathEffect;
/**
 * Maps a CSS mix-blend-mode value to the corresponding Canvas {@link GlobalCompositeOperation}.
 * All CSS blend mode values are supported by the Canvas 2D API except 'normal',
 * which maps to the default 'source-over'.
 */
const MIX_BLEND_MODE_TO_COMPOSITE = {
    [mix_blend_mode_1.MIX_BLEND_MODE.NORMAL]: 'source-over',
    [mix_blend_mode_1.MIX_BLEND_MODE.MULTIPLY]: 'multiply',
    [mix_blend_mode_1.MIX_BLEND_MODE.SCREEN]: 'screen',
    [mix_blend_mode_1.MIX_BLEND_MODE.OVERLAY]: 'overlay',
    [mix_blend_mode_1.MIX_BLEND_MODE.DARKEN]: 'darken',
    [mix_blend_mode_1.MIX_BLEND_MODE.LIGHTEN]: 'lighten',
    [mix_blend_mode_1.MIX_BLEND_MODE.COLOR_DODGE]: 'color-dodge',
    [mix_blend_mode_1.MIX_BLEND_MODE.COLOR_BURN]: 'color-burn',
    [mix_blend_mode_1.MIX_BLEND_MODE.HARD_LIGHT]: 'hard-light',
    [mix_blend_mode_1.MIX_BLEND_MODE.SOFT_LIGHT]: 'soft-light',
    [mix_blend_mode_1.MIX_BLEND_MODE.DIFFERENCE]: 'difference',
    [mix_blend_mode_1.MIX_BLEND_MODE.EXCLUSION]: 'exclusion',
    [mix_blend_mode_1.MIX_BLEND_MODE.HUE]: 'hue',
    [mix_blend_mode_1.MIX_BLEND_MODE.SATURATION]: 'saturation',
    [mix_blend_mode_1.MIX_BLEND_MODE.COLOR]: 'color',
    [mix_blend_mode_1.MIX_BLEND_MODE.LUMINOSITY]: 'luminosity'
};
class BlendEffect {
    constructor(mixBlendMode) {
        this.mixBlendMode = mixBlendMode;
        this.type = 4 /* EffectType.BLEND */;
        this.target = 2 /* EffectTarget.BACKGROUND_BORDERS */ | 4 /* EffectTarget.CONTENT */;
        this.compositeOperation = MIX_BLEND_MODE_TO_COMPOSITE[mixBlendMode];
    }
}
exports.BlendEffect = BlendEffect;
class FilterEffect {
    constructor(filterString) {
        this.filterString = filterString;
        this.type = 5 /* EffectType.FILTER */;
        this.target = 2 /* EffectTarget.BACKGROUND_BORDERS */ | 4 /* EffectTarget.CONTENT */;
    }
}
exports.FilterEffect = FilterEffect;
const isTransformEffect = (effect) => effect.type === 0 /* EffectType.TRANSFORM */;
exports.isTransformEffect = isTransformEffect;
const isClipEffect = (effect) => effect.type === 1 /* EffectType.CLIP */;
exports.isClipEffect = isClipEffect;
const isOpacityEffect = (effect) => effect.type === 2 /* EffectType.OPACITY */;
exports.isOpacityEffect = isOpacityEffect;
const isClipPathEffect = (effect) => effect.type === 3 /* EffectType.CLIP_PATH */;
exports.isClipPathEffect = isClipPathEffect;
const isBlendEffect = (effect) => effect.type === 4 /* EffectType.BLEND */;
exports.isBlendEffect = isBlendEffect;
const isFilterEffect = (effect) => effect.type === 5 /* EffectType.FILTER */;
exports.isFilterEffect = isFilterEffect;
